comms/dht/src/dedup/dedup_cache.rs

```diff
@@ -72,96 +70,91 @@ impl DedupCacheDatabase {
-                    .values((
-                        dedup_cache::body_hash.eq(&body_hash),
-                        dedup_cache::sender_public_key.eq(&public_key),
-                        dedup_cache::number_of_hits.eq(1),
-                        dedup_cache::last_hit_at.eq(Utc::now().naive_utc()),
-                    ))
-                    .execute(conn);
-                match insert_result {
-                    Ok(1) => Ok(1),
-                    Ok(n) => Err(StorageError::UnexpectedResult(format!(
-                        "Expected exactly one row to be inserted. Got {}",
-                        n
-                    ))),
-                    Err(diesel::result::Error::DatabaseError(kind, e_info)) => match kind {
-                        DatabaseErrorKind::UniqueViolation => {
-                            // Update hit stats for the message
-                            diesel::update(dedup_cache::table.filter(dedup_cache::body_hash.eq(&body_hash)))
-                                .set((
-                                    dedup_cache::sender_public_key.eq(&public_key),
-                                    dedup_cache::number_of_hits.eq(dedup_cache::number_of_hits + 1),
-                                    dedup_cache::last_hit_at.eq(Utc::now().naive_utc()),
-                                ))
-                                .execute(conn)?;
-                            // TODO: Diesel support for RETURNING statements would remove this query, but is not
-                            //       available for Diesel + SQLite yet
-                            let hits = dedup_cache::table
-                                .select(dedup_cache::number_of_hits)
-                                .filter(dedup_cache::body_hash.eq(&body_hash))
-                                .get_result::<i32>(conn)?;
+    fn insert_body_hash_or_update_stats(&self, body_hash: String, public_key: String) -> Result<u32, StorageError> {
+        let insert_result = {
+            let conn = self.connection.get_pooled_connection()?;
+            diesel::insert_into(dedup_cache::table)
+                .values((
+                    dedup_cache::body_hash.eq(&body_hash),
+                    dedup_cache::sender_public_key.eq(&public_key),
+                    dedup_cache::number_of_hits.eq(1),
+                    dedup_cache::last_hit_at.eq(Utc::now().naive_utc()),
+                ))
+                .execute(&conn)
+        };
+        match insert_result {
+            Ok(1) => Ok(1),
+            Ok(n) => Err(StorageError::UnexpectedResult(format!(
+                "Expected exactly one row to be inserted. Got {}",
+                n
+            ))),
+            Err(diesel::result::Error::DatabaseError(kind, e_info)) => match kind {
+                DatabaseErrorKind::UniqueViolation => {
+                    // Update hit stats for the message
+                    {
+                        let conn = self.connection.get_pooled_connection()?;
```

Comment:
Why not get the pool connection once in this function?

Reply:
The pooled connection is held until the code goes out of scope